### PR TITLE
Clean software cron

### PR DESCRIPTION
**Pull request by AdrienClairembault**
Internal ref: 18803.

Clean: 
- Software versions without installations or linked licenses
- Software without versions

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

**Comment by AdrienClairembault** [comment]
>     1. You should add a cronInfo method to be able to translate cron description.
> 
>     2. You should add a parameter to limit volume on each execution (see mailgate or queuednotification crons to get usage example).
> 
>     3. Isolating cron into a dedicated class is a good idea. However, as cron was initally made to be attached to an itemtype (which means a CommonDBTM class to me), can you check that all works as excpected (i.e. setting cron settings, display of cron logs, ...) ?

All points should be OK now.
For your third point, this dedicated class is in fact a `CommonDBTM` (extends) because, as you said, it's meant to work this way :/.

IMO, in a future version, all crontask should be extracted, put in a common namespace and follow a distinct interface which is not related to CommonDBTM (same should be done for massive actions too).

**Comment by AdrienClairembault** [comment]
Not sure what happened on circle CI (build canceled somehow ?) but tests on github actions are OK.

File: inc/cleansoftwarecron.class.php [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+      );
+
+      // Move software with no versions in the thrashbin
+      $total += $task->addVolume(
+         self::deleteItems(
+            self::getSoftwareWithNoVersions(),
+            new Software(),
+            $max - $total
+         )
+      );
```
Comment (cedric-anne) [inline inc/cleansoftwarecron.class.php:73]:
```suggestion
      self::deleteItems(
         self::getSoftwareWithNoVersions(),
         new Software(),
         $max - $total
      );
```

File: inc/cleansoftwarecron.class.php [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+      return 1;
+   }
+
+   /**
+    * Get all software versions which are not installed
+    *
+    * @return DBmysqlIterator
+    */
+   protected static function getVersionsWithNoInstallation(): array {
```
Comment (cedric-anne) [inline inc/cleansoftwarecron.class.php:85]: 1. PHPDoc is no more valid.
2. Maybe add `Criteria` to method name (this is generally a word contained by methods returning DB query "descriptor").

File: inc/cleansoftwarecron.class.php [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+               ],
+            ],
+         ],
+      ];
+   }
+
+   /**
+    * Get all software with no versions
+    */
+   protected static function getSoftwareWithNoVersions(): array {
```
Comment (cedric-anne) [inline inc/cleansoftwarecron.class.php:119]: Same remark.

File: inc/cleansoftwarecron.class.php [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+      CommonDBTM $em,
+      int $max
+   ): int {
+      global $DB;
+
+      $total = 0;
+
+      do {
+         $scope['LIMIT'] = min($max - $total, self::MAX_BATCH_SIZE);
+         $items = $DB->request(self::getVersionsWithNoInstallation($scope));
```
Comment (cedric-anne) [inline inc/cleansoftwarecron.class.php:155]:
```suggestion
         $items = $DB->request($scope);
```

File: inc/cleansoftwarecron.class.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+ * ---------------------------------------------------------------------
+ */
+
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access this file directly");
+}
+
+class CleanSoftwareCron extends CommonDBTM
+{
+   const task_name = 'cleansoftware';
```
Comment (cedric-anne) [inline inc/cleansoftwarecron.class.php:39]:
```suggestion
   const TASK_NAME = 'cleansoftware';
```

File: install/empty_data.php [hunk 6]
```diff
@@ -666,6 +666,16 @@
       'mode'          => 2,
       'lastrun'       => null,
       'logs_lifetime' => 30,
+   ], [
+      'id'            => 38,
+      'itemtype'      => CleanSoftwareCron::getType(),
+      'name'          => CleanSoftwareCron::task_name,
```
Comment (cedric-anne) [inline install/empty_data.php:672]:
```suggestion
      'name'          => CleanSoftwareCron::TASK_NAME,
```

File: install/update_94_95.php [hunk 7]
```diff
@@ -2000,6 +2000,25 @@ function update94to95() {
    }
    // /use_kerberos
 
+   // CleanSoftwareCron cron task
+   $cron_em = new CronTask();
+   $tasks = $cron_em->find(['name' => CleanSoftwareCron::task_name]);
```
Comment (cedric-anne) [inline install/update_94_95.php:2005]:
```suggestion
   $tasks = $cron_em->find(['name' => CleanSoftwareCron::TASK_NAME]);
```

File: install/empty_data.php [hunk 8]
```diff
@@ -666,6 +666,16 @@
...
+      'frequency'     => 60 * 60 * 24 * 30,
```
Comment (cedric-anne) [inline install/empty_data.php:673]:
```suggestion
      'frequency'     => MONTH_TIMESTAMP,
```

File: install/update_94_95.php [hunk 9]
```diff
@@ -2000,6 +2000,25 @@ function update94to95() {
...
+
+   // Task already exist
+   if (count($tasks) === 0) {
+      $cron_em->add([
+         'itemtype'      => CleanSoftwareCron::class,
+         'name'          => CleanSoftwareCron::task_name,
```
Comment (cedric-anne) [inline install/update_94_95.php:2011]:
```suggestion
         'name'          => CleanSoftwareCron::TASK_NAME,
```

File: install/update_94_95.php [hunk 10]
```diff
@@ -2000,6 +2000,25 @@ function update94to95() {
...
+         'frequency'     => 60 * 60 * 24 * 30, // Once per month
```
Comment (cedric-anne) [inline install/update_94_95.php:2012]:
```suggestion
         'frequency'     => MONTH_TIMESTAMP,
```

File: inc/cleansoftwarecron.class.php [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+   const MAX_BATCH_SIZE = 2000;
+
```
Comment (cedric-anne) [inline inc/cleansoftwarecron.class.php:42]:
```suggestion

   static protected $notable = true;

```

File: inc/cleansoftwarecron.class.php [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+         // Stop if no items found
+      } while ($count > 0);
+
+      return $total;
+   }
+
+   public function isEntityAssign() {
+      return false;
+   }
```
Comment (cedric-anne) [inline inc/cleansoftwarecron.class.php:171]: Why is it required ?

File: inc/cleansoftwarecron.class.php (hunk 11 above)
Comment (AdrienClairembault) [reply inc/cleansoftwarecron.class.php:42]: Was not aware this existed thanks :)

File: inc/cleansoftwarecron.class.php (hunk 12 above)
Comment (AdrienClairembault) [reply inc/cleansoftwarecron.class.php]: Not needed anymore with $notable

**Comment by AdrienClairembault** [comment]
Thanks for the great review @cedric-anne.
Not sure how I messed up so many things when moving code around last thursday :D

Everything should be good now :)

**Comment by AdrienClairembault** [comment]
Maybe we can use a temporary name like update_95_nextmajor ?
This way we can just rename it before the release.

**Comment by trasher** [comment]
You can let the PR as is for now; we're aware it must be done ;)

**Comment by cedric-anne** [comment]
> As it will not be merged on 9.5, update should be done on a new file, but we did not yet decided the number of the next version, so you will have to wait to finish this PR.

I rebased on master and moved migration instructions.

Can be reviewed and merged.

**Comment by orthagh** [comment]
Do you think a relative cli command is a good idea ?

**Comment by trasher** [comment]
> Do you think a relative cli command is a good idea ?

It may be, be this can be done on another PR

File: inc/console/assets/cleansoftwarecommand.class.php [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+      parent::configure();
+
+      $this->setName('glpi:assets:cleansoftware');
+      $this->setAliases(['assets:cleansoftware']);
+      $this->setDescription(CleanSoftwareCron::getTaskDescription());
+
+      $this->addOption(
+         'max',
+         'm',
+         InputOption::VALUE_OPTIONAL,
```
Comment (cedric-anne) [inline inc/console/assets/cleansoftwarecommand.class.php:58]: Option is not required, but if passed to the command, its value is required.
```suggestion
         InputOption::VALUE_REQUIRED,
```

**Review by cedric-anne — APPROVED** [review]

File: install/update_95_xx/softwares.php [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+// CleanSoftwareCron cron task
+$cron_em = new CronTask();
```
Comment (trasher) [inline install/update_95_xx/softwares.php:34]: Maybe just use `Crontab::register()`

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Comment by vollkommenIrrelevant** [comment]
Great news!

**Pull request merged**